tsdb/store.go

```diff
@@ -534,28 +534,67 @@ func (s *Store) DeleteShard(shardID uint64) error {
 		return nil
 	}
 
-	// Remove the shard from the database indexes before closing the shard.
-	// Closing the shard will do this as well, but it will unload it while
-	// the shard is locked which can block stats collection and other calls.
-	sh.UnloadIndex()
+	// Remove the shard from Store so it's not returned to callers requesting
+	// shards.
+	s.mu.Lock()
+	delete(s.shards, shardID)
+	s.mu.Unlock()
 
-	if err := sh.Close(); err != nil {
+	// Get the shard's local bitset of series IDs.
+	index, err := sh.Index()
+	if err != nil {
 		return err
 	}
 
-	if err := os.RemoveAll(sh.path); err != nil {
-		return err
+	var ss *SeriesIDSet
+	if i, ok := index.(interface {
+		SeriesIDSet() *SeriesIDSet
+	}); ok {
+		ss = i.SeriesIDSet()
 	}
 
-	if err := os.RemoveAll(sh.walPath); err != nil {
+	db := sh.Database()
+	if err := sh.Close(); err != nil {
 		return err
 	}
 
-	s.mu.Lock()
-	delete(s.shards, shardID)
-	s.mu.Unlock()
+	// Determine if the shard contained any series that are not present in any
+	// other shards in the database.
+	shards := s.filterShards(byDatabase(db))
 
-	return nil
+	s.walkShards(shards, func(sh *Shard) error {
+		index, err := sh.Index()
+		if err != nil {
+			return err
+		}
+
+		if i, ok := index.(interface {
+			SeriesIDSet() *SeriesIDSet
+		}); ok {
+			ss.Diff(i.SeriesIDSet())
```

Comment:
Does this return a new bitset or mutate the existing one?

Reply:
It mutates the existing one, making it slightly different to the `AndNot` method in the series set.